Watch improvements

eatkins opened the pull request:
This pull request makes some improvements to the source modification watch.

- The PollingWatchService had a few minor issues that caused transient test failures
- The MacOSXWatchService did not quite implement the WatchService api correctly (DefaultWatchServiceSpec failed with the watch service set to MacOSXWatchService
- I re-implement the SourceModificationWatch.watch method to remove the latency between receiving file or user input events and triggering a build (or exiting).

eatkins committed:
Apply recursive parameter in source filter

Right now, even with the recursive toggle, the Source.accept method will
accept files that are not in the base path.

I also added a simple test suite for Source to verify the basic
semantics.

eatkins committed:
Reduce watch service latency

There is often noticeable lag between when a file event happens and when
the build is triggered. There is also often latency when the user
presses enter to interrupt the watch. Both are due to the way that
SourceModificationWatch polls for results. It can be removed by having
the main thread block on an event queue that is asynchronously filled by
threads that wait for user input and file events. I made a similar
change in CloseWatch.

The new EventMonitor is where most of the logic lives. It is implemented
as a block box that effectively provides three methods:
1) watch() -- block the main thread until a user input or watched file
   event occurs, returning true if the build was triggered
2) state() -- a snapshot of the current watch state. This is primarily
   for legacy compatibility with sbt, which has a number of methods that
   take the state as input. In practice, all that sbt really needs most
   of the time is the count of the number of build triggers
3) close() -- shutdown any threads or service started up by the event
   monitor

I implemented the EventMonitor as a block box so that it would be
straightforward to change the existing implementation or add new
implementations without having to break forward and backwards
compatibility with sbt. In particular, I can envision adding a second
EventMonitor that uses a persistent file system cache for file events
instead of the ad-hoc cache that currently exists in the
EventMonitorImpl.eventThread.

At the moment, the one implementation is EventMonitorImpl. It spins up
two threads for monitoring user input and the file events. Both write
`EventMonitor.Events` to a concurrent queue that the main thread reads
from. User input events supersede file triggers. When the user hits
enter, the queue is cleared and filled with the exit event. Otherwise,
the event thread polls the watch service for events. When it receives a
file event, it adds the event to a cache of recent events. This cache is
used to prevent multiple builds from being triggered by the same file*.
The eventThread also detects when directories are created or deleted and
registers or unregisters the directory with the watch service as needed.

I also stopped registering all of the files. Only directories are
registered now. The registered files just waste memory.

This commit also adds logging of watch events. Unfortunately, I can't
get the logging to actually work at the debug level due to
https://github.com/sbt/sbt/issues/4097, but once that issue is fixed,
logging with the new EventMonitor should work.

I added a simple test that the anti-entropy works and made some small
adjustments to the existing tests to make them work with the new
implementation.

For backwards compatibility with older versions of sbt, I re-implement
SourceModificationWatch.watch to wrap an EventMonitor that we shutdown
after each trigger. This is also the reason that I had to add two close
methods to EventMonitor -- one that shuts down the watch service, and
one that doesn't.

* When neovim, for example, saves a file. It moves the buffer into the
file location which can trigger a delete and a create event on the file.
Without the anti-entropy timeout, there would be two build triggered for
the same file save.

eatkins committed:
Bump appleFileEvents

I discovered an issue in the apple-file-events library that could cause
certain api calls to induce a segfault. I fixed the bug upstream:
https://github.com/swoval/swoval/pull/11/commits/78d9c03f4eeb7d8f0b086edb08999ae61e8d7d6f

eatkins committed:
Use the MacOSXWatchService in DefaultWatchServiceSpec

It was an oversight that I didn't make this change when adding the
MacOSXWatchService.

eatkins committed:
Set file times on register

When registering a new path, we should not trigger events for the files
that already exist in the directory. To fix this, we calculate the last
modified times for all of the files in the newly registered path and add
them to the list.

This fixed spurious test failures that I was having in my anti-entropy
spec.

eatkins committed:
Add unregister to WatchService

When a watched directroy is deleted, it's good practice to cleanup the
resources associated with it. If it's recreated, we can just
re-register. For binary compatibility reasons, I couldn't add unregister
to WatchService, so I added a trait that gets mixed into all of the
implementations. The method is then accessible in abstract WatchServices
by runtime pattern matching.

This change also has the side effect of preventing the default
WatchService from creating redundant registrations.

Bonus: fix some compiler warnings

eatkins committed:
Use realPath to register in MacOSXWatchService

Temporary directories on osx generally have a different value for path
vs. realPath (the latter usually has a leading /private). The events
from the apple file system api are real paths, with the leading
/private, so in order to check if the event is for a registered path, we
must register using the real path.

eatkins committed:
Monitor streams

The apple file system api only offers recursive directory listeners,
which makes it pointless to have a listener for each subdirectory.
Whenever a stream is registered with FileEventsApi and there are
previously registered streams that are subdirectories of this stream,
the FileEventsApi removes the now redundant streams and fires a callback
with the name of each directory whose events are now triggered by the
parent stream. When this callback is fired, the streamId associated with
the stream is no longer valid. In this commit, I start tracking the
registered streams and don't bother registering a subdirectory with the
watch service if one of its parents is already being monitored.

eatkins committed:
Fire events for base paths

Prior to this commit, if a file event was detected for a base directory,
no event was fired because the parent of the base directory wasn't
monitored.

Bonus: check if the service is open in a few methods and throw a
ClosedWatchServiceException if it isn't.

eatkins committed:
Defer removal of deleted files in PollingWatchService

I noticed that when a directory was removed, the PollingWatchService
didn't always report file events. This was because sometimes the parent
directory was removed from the `watched` map before the delete file was
handled in the loop. This caused the loop to treat it like an
unmonitored file.

eatkins committed:
Reduce timeouts

These timeouts are all excessively long from what I can tell. This
change makes the whole test suite take ~10 seconds instead of ~45
seconds (the *WatchServiceSpec decline from ~45 seconds to ~2.5 seconds
on my machine). I found, however, that on travis, longer timeouts are
needed because the jvm needs to warm up and the timeouts are not short
enough in that scenario.

eatkins committed:
Use real path in SourceModificationWatchSpec

The SourceModificationWatchSpecs fail on OSX because temp directories
are reported by the watch service with an additional prefix "/private".
This prefix is added when toRealPath is called on the temp directory.

eatkins committed:
Make Source fields public

It is a pain for plugin writers that these fields are private. I had to
use reflection to extract them in CloseWatch.

typesafe-tools commented:
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=1.1.x&zinc=1.1.x&io=pull/142/head&librarymanagement=1.1.x&util=1.1.x&website=1.1.x&deploy_to_bintray=false&run_sbt_community=false&verbose_dbuild=false&community_verbose=false">this link.</A>

eed3si9n commented:
@eatkins Thanks for the update.
 I am going to be busy this week and next for conference, so the review might delay.

eatkins commented:
Sounds good. I have some improvements I want to make anyway.

He-Pin commented:
https://github.com/facebook/watchman

eatkins commented:
Care to elaborate, @hepin1989?

dwijnand commented on io/src/main/scala/sbt/internal/io/SourceModificationWatch.scala:
```diff
@@ -156,7 +156,8 @@ final class Source(
 
-    p.startsWith(base.toPath) && inc.accept(p.toFile) && !excludeFilter.accept(p.toFile)
+    (if (!recursive) p.getParent == base.toPath else p.startsWith(base.toPath)) && inc.accept(
+      p.toFile) && !excludeFilter.accept(p.toFile)
```
if could we add a test for this oversight, it would be nice.

eatkins replied:
Done.

dwijnand commented on io/src/main/scala/sbt/internal/io/SourceModificationWatch.scala:
```diff
@@ -136,10 +136,10 @@ private[sbt] final class WatchState private (
+    val base: File,
+    val includeFilter: FileFilter,
+    val excludeFilter: FileFilter,
+    val recursive: Boolean
```
I agree! https://github.com/sbt/io/pull/123 xD

dwijnand reviewed (dismissed):
despite its size reviewing this was a real pleasure commit-by-commit.. until the last commit bomb :)

I don't have the bandwidth to review it in detail, but I like the sound of it.

I'll give Eugene a chance to weigh in in case he weighs in differently to me.

eatkins commented:
Thanks @dwijnand. Yeah, that last commit was hard to split up into bite sized chunks. Apologies for the review overhead.

eed3si9n commented on io/src/main/scala/sbt/internal/io/EventMonitor.scala:
```diff
@@ -0,0 +1,194 @@
+import scala.collection.JavaConverters._
+import scala.concurrent.duration._
+
+private[sbt] sealed trait EventMonitor extends AutoCloseable {
```
Could you add the overview of what EventMonitor is as ScalaDoc, plz? (Something similar to what you wrote in commit comment is fine)

eed3si9n commented on io/src/main/scala/sbt/internal/io/EventMonitor.scala:
```diff
@@ -0,0 +1,194 @@
+  def state(): WatchState
+
+  /** Block indefinitely until the monitor receives a file event or the user stops the watch. */
+  def watch(): Boolean
```
I wonder if there might be a better blocky name, like `poll()` or `awaitEvent()`.

eed3si9n reviewed (changes requested):
Overall looks good, but I left some minor comments.

I am thankful to some of the macOS specific changes, yet I am nervous about them as well since they are not currently tested on Travis CI. Could you add macOS to test matrix following https://github.com/sbt/sbt-launcher-package/pull/219/commits/66b8670c59c425e309fc932cd4e214841a6ca641?

eatkins committed:
Add osx to the travis build

Modified from: https://github.com/sbt/sbt-launcher-package/commit/66b8670c59c425e309fc932cd4e214841a6ca641

eatkins commented:
Adding mac os to travis seems like a no brainer. I'm working on the matrix, but travis seems to be having issues at the moment.

eatkins commented:
Figured out Travis. Somewhat frustratingly, I'm having trouble getting the PollingWatchServiceSpec to pass on osx (but the DefaultWatchServiceSpec works pretty reliably well). At any rate, I'm sure I'll get the tests working tomorrow.

dwijnand commented on io/src/test/scala/sbt/internal/io/DefaultWatchServiceSpec.scala:
```diff
@@ -8,14 +8,12 @@ import scala.concurrent.duration._
-  // We give it more time to detect changes.
-  val (pollDelay, maxWaitTime) = (50.milliseconds, 3.seconds)
+  val pollDelay =
+    Option(System.getProperty("io.watch.poll.ms")).fold(50.milliseconds)(_.toInt.milliseconds)
```
could you prefix the prop key with `sbt.` please?

eatkins replied:
I was only using this for travis debugging. It's gone now. I will keep in mind prefixing keys with sbt. if it comes up again though.

eatkins committed:
Make PollingThread.start block until startup

Rather than spinning on PollingThread.initDone in the init() method, I
override the start() method of PollingThread to block until the created
thread is done initializing. This should reduce the average runtime of
init() by 50ms in most cases. I am sure that it doesn't matter, but I
also didn't like reading from thread.initDone directly without
synchronization from a different thread.

For binary compatibility had to provide a getter and setter for
_initDone (each of which I deprecated). I suspect that no one is using
these apis outside of the PollingWatchService file since PollingThread
is private which suggests that it might make sense to just add a mima
filter instead.

eatkins committed:
Set new file modified times in test

I was having a lot of trouble getting the PollingWatchServiceSpec to
pass on travis with osx and it turns out this is because the osx agents
are using HFS+ not APFS. The HFS+ file system doesn't support
millisecond precision on modified times. To fix this, set the file
modified time to be in the past when writing a new file.

eatkins committed:
Don't directly expose keys

The keysWithEvents was exposing unsynchronized mutable data. Even though
it was only accessed from a different thread inside of synchronized
blocks, it seems better to guarantee this behavior by adding a helper
accessor that ensures synchronization.

For binary compatibility had to provide a getter (that I deprecated) for
_keysWithEvents. I suspect that no one is using these apis outside of
the PollingWatchService file since PollingThread is private which
suggests that it might make sense to just add a mima filter instead.

eatkins committed:
Fix implicit numeric widening error message

eatkins commented:
I figured out the travis issues with PollingWatchServiceSpec (among other things, the travis osx agents use HFS+, which doesn't have millisecond precision). I also added the scaladocs to EventMonitor as requested. Finally, I renamed `EventMonitor.watch` to `EventMonitor.awaitEvents` as suggested by @eed3si9n.

eed3si9n reviewed (dismissed):
LGTM

eatkins committed:
Retry setModifiedTimeOrFalse on error

Occasionally on Travis, I get errors where IO.setModifiedTimeOrFalse
throws an exception because utimensat returns an EINVAL. Fixing whatever
is going on in IO.setModifiedTimeOrFalse is beyond the scope of my work,
but I want the travis builds to not be flaky for now.

dwijnand reviewed (approved):
LGTM. Thanks @eatkins!

cunei commented:
@eatkins Can you please provide a link to a failing Travis run in which you see `utimensat()` returning `EINVAL`? According to the man page, there are only a few cases in which that may happen, the most likely being a NULL path name.

eatkins commented:
@cunei https://travis-ci.org/sbt/io/jobs/370883894
